src/ts_catalog/chunk_data_node.c

```diff
@@ -312,34 +332,53 @@ chunk_data_node_tuple_delete(TupleInfo *ti, void *data)
 int
 ts_chunk_data_node_delete_by_chunk_id(int32 chunk_id)
 {
+	ScanTupLock tuplock = {
+		.lockmode = LockTupleExclusive,
+		.waitpolicy = LockWaitBlock,
+	};
```

Comment:
Taking this lock here won't do anything. A LockTupleExclusive is implicitly taken when doing a delete.

If you already did `FOR UPDATE`, then no extra locking should be needed here.

Another option is to avoid `FOR UPDATE` locks and simply check the delete lock result. If it returns `TM_Deleted`, then the tuple was already deleted by someone else and we need to do nothing.

The issue, however, that the CatalogDelete API we are using doesn't expose the lock/delete result. So, we can either call heap_delete_tuple ourselves or proactively take a lock here and check the lock result in the tuple handling function (`chunk_data_node_tuple_delete`).